pjmedia/src/pjmedia-codec/and_vid_mediacodec.cpp

```diff
@@ -0,0 +1,1846 @@
+                                      pj_size_t count,
+                                      pjmedia_frame packets[],
+                                      unsigned out_size,
+                                      pjmedia_frame *output)
+{
+    struct and_media_codec_data *and_media_data;
+    pj_status_t status = PJ_EINVAL;
+
+    PJ_ASSERT_RETURN(codec && count && packets && out_size && output,
+                     PJ_EINVAL);
+    PJ_ASSERT_RETURN(output->buf, PJ_EINVAL);
+
+    and_media_data = (and_media_codec_data*) codec->codec_data;
+    and_media_data->dec_has_output_frame = PJ_FALSE;
+    and_media_data->dec_input_buf = NULL;
+    and_media_data->dec_input_buf_len = 0;
+
+    if (and_media_codec[and_media_data->codec_idx].decode) {
+        status = and_media_codec[and_media_data->codec_idx].decode(codec, count,
+                                                           packets, out_size,
+                                                           output);
+    }
+    if (status != PJ_SUCCESS) {
+	return status;
+    }
+    if (!and_media_data->dec_has_output_frame) {
+	pjmedia_event event;
+
+	/* Broadcast missing keyframe event */
+	pjmedia_event_init(&event, PJMEDIA_EVENT_KEYFRAME_MISSING,
+			   &packets[0].timestamp, codec);
+	pjmedia_event_publish(NULL, codec, &event,
+			      PJMEDIA_EVENT_PUBLISH_DEFAULT);
+
+	PJ_LOG(4,(THIS_FILE, "Decoder couldn't produce output frame"));
+
+	output->type = PJMEDIA_FRAME_TYPE_NONE;
+	output->size = 0;
+	output->timestamp = packets[0].timestamp;
+    }
+    return PJ_SUCCESS;
+}
+
+#if PJMEDIA_HAS_AND_MEDIA_H264
+
+static pj_status_t open_h264(and_media_codec_data *and_media_data)
+{
+    pj_status_t status;
+    pjmedia_vid_codec_param *param = and_media_data->prm;
+    pjmedia_h264_packetizer_cfg  pktz_cfg;
+    pjmedia_vid_codec_h264_fmtp  h264_fmtp;
+    h264_codec_data *h264_data;
+
+    /* Parse remote fmtp */
+    pj_bzero(&h264_fmtp, sizeof(h264_fmtp));
+    status = pjmedia_vid_codec_h264_parse_fmtp(&param->enc_fmtp,
+                                               &h264_fmtp);
+    if (status != PJ_SUCCESS)
+        return status;
+
+    /* Apply SDP fmtp to format in codec param */
+    if (!param->ignore_fmtp) {
+        status = pjmedia_vid_codec_h264_apply_fmtp(param);
+        if (status != PJ_SUCCESS)
+            return status;
+    }
+    h264_data = PJ_POOL_ZALLOC_T(and_media_data->pool, h264_codec_data);
+    if (!h264_data)
+        return PJ_ENOMEM;
+
+    pj_bzero(&pktz_cfg, sizeof(pktz_cfg));
+    pktz_cfg.mtu = param->enc_mtu;
+    pktz_cfg.unpack_nal_start = 4;
+    /* Packetization mode */
+    if (h264_fmtp.packetization_mode == 0)
+        pktz_cfg.mode = PJMEDIA_H264_PACKETIZER_MODE_SINGLE_NAL;
+    else if (h264_fmtp.packetization_mode == 1)
+        pktz_cfg.mode = PJMEDIA_H264_PACKETIZER_MODE_NON_INTERLEAVED;
+    else
+        return PJ_ENOTSUP;
+
+    pktz_cfg.mode = PJMEDIA_H264_PACKETIZER_MODE_NON_INTERLEAVED;
```

Comment:
Looks like we need to mention somewhere in the docs that Android H264 only support packetization mode non-interleaved. Also just checked vid_toolbox impl, it seems to have the same issue.

Ideally the codec implementation can verify H264 fmtp setting from `pjmedia_codec_mgr_set_default_param()` to make sure that SDP fmtp `packetization-mode` is set to 1. Unfortunately this cannot be done for now.

Note that when the SDP nego result uses `packetization mode=0` (single-nal) and we keep sending/encoding with non-interleaved RTP packetization, remote will not be able to unpack our RTP payload and it won't be able to decode our video properly.